pymbolic/mapper/flattener.py

```diff
@@ -23,18 +31,75 @@
 THE SOFTWARE.
 """
 
+
+import pymbolic.primitives as p
 from pymbolic.mapper import IdentityMapper
+from pymbolic.typing import ExpressionT
+
 
+class FlattenMapper(IdentityMapper[[]]):
```

Comment:
@alexfikl This PR bumps the compat target to 3.10. The immediate reason was this inheritance, which is illegal in 3.9 and below. It can be hacked around, but it would have to be hacked around in a lot of places. 3.8 is EOL as of a few weeks ago, and I'm not sure I'm too enamored with supporting 3.9. Current numpy also has a 3.10 minimum. The net implication is that the whole stack will probably switch before too long. There's some ruff type annotation churn that results, but it's probably for the best to get that out of the way. How do you feel about that idea?

Reply:
That sounds good to me. I've also jumped directly to 3.10 with most of my random code due to numpy bumping it.

Reply:
Also, I'm not sure I've ever seen this before. What's `IdentityMapper[[]]` supposed to mean?

Reply:
The corresponding type parameter on `IdentityMapper` is a `ParamSpec` for i.e. the extra arguments that get passed around by the mapper methods.

https://docs.python.org/3.13/library/typing.html#user-defined-generic-types says

> The one exception to this is that a list of types can be used to substitute a [ParamSpec](https://docs.python.org/3.13/library/typing.html#typing.ParamSpec)

Based on the failures, this must be a relatively recent thing.